packages/protocol/contracts/L1/StakingProverPool3.sol

```diff
@@ -0,0 +1,316 @@
+            // provers[31].stakedAmount == 0
+            unchecked {
+                ++currentIdIdx;
+            }
+
+            // If we just filled up the last place in the array (so means full -
+            // we can
+            // determine which is the staker in risk)
+            if (provers[31].stakedAmount != 0) {
+                (minimumStakerId, minimumStake) = _determineLowestStaker();
+            }
+        } else {
+            // List is full, we need to check which position we need to put them
+            // into
+            (minimumStakerId,) = _determineLowestStaker();
+
+            // We need to overwrite it's place
+            stakers[minimumStakerId] = Staker(
+                minimumStakerId, prover, totalAmount, rewardPerGas, capacity
+            );
+            provers[minimumStakerId] = Prover(totalAmount, rewardPerGas);
+            proverToId[prover] = minimumStakerId;
+
+            // Need to determine again who is the last / lowest staker
+            (minimumStakerId, minimumStake) = _determineLowestStaker();
+        }
+
+        TaikoToken(AddressResolver(this).resolve("taiko_token", false)).burn(
+            prover, totalAmount
+        );
+
+        emit ProverEntered(prover, totalAmount, rewardPerGas, capacity);
+    }
+
+    function stakePositionAdjustments(
+        uint32 newTotalAmount,
+        address prover, // This can be a 'delegate' so not using msg.sender
+        uint16 rewardPerGas,
+        uint16 capacity
+    )
+        external
+        onlyProver(prover)
+    {
+        // Cannot enter the pool below the minimum - we are not allowing the
+        // minium staker to lower it's position
+        require(minimumStake < newTotalAmount, "Cannot enter below minimum");
+        require(rewardPerGas > 0, "Cannot be less than 1");
+
+        Staker memory mStaker = stakers[proverToId[prover]];
+
+        if (mStaker.stakedAmount > newTotalAmount) {
+            // Lowered position, so basically we just need to burn some tokens
+            // for him/her
+            TaikoToken(AddressResolver(this).resolve("taiko_token", false)).mint(
+                prover, mStaker.stakedAmount - newTotalAmount
+            );
+        } else if (mStaker.stakedAmount < newTotalAmount) {
+            // Otherwise it raised it's stake, so burn more tokens from it's
+            // balance
+            TaikoToken(AddressResolver(this).resolve("taiko_token", false)).burn(
+                prover, newTotalAmount - mStaker.stakedAmount
+            );
+        }
+
+        stakers[proverToId[prover]] = Staker(
+            minimumStakerId, prover, newTotalAmount, rewardPerGas, capacity
+        );
+        provers[proverToId[prover]] = Prover(newTotalAmount, rewardPerGas);
+
+        // Need to determine again who is the last / lowest staker
+        (minimumStakerId, minimumStake) = _determineLowestStaker();
+
+        emit ProverModified(prover, newTotalAmount, rewardPerGas, capacity);
```

Comment:
IMO, the original design of the API was much better, as this design relies on the user providing their old values as defaults to this all-encompassing function, rather than the other design only providing the value you want to update? Each function would also be much less gas intensive, IE I'm paying for a lot of gas to do other stuff when I just want to adjust capacity?